### PR TITLE
core: refuse dbus activation if dbus is not running

### PR DESCRIPTION
**Pull request by YHNdnzj**
dbus-broker issues `StartUnit` directly for activation requests, so let's add a check on bus state in `bus_unit_queue_job` to refuse that if dbus is not running.

Replaces https://github.com/systemd/systemd/pull/27570
Closes https://github.com/systemd/systemd/issues/26799

### CODE REVIEW

File: src/core/manager.h [hunk 1]
```diff
@@ -504,6 +504,12 @@ DEFINE_TRIVIAL_CLEANUP_FUNC(Manager*, manager_free);
 
 int manager_startup(Manager *m, FILE *serialization, FDSet *fds, const char *root);
 
+bool manager_dbus_is_running_full(Manager *m, bool deserialized);
+static inline bool manager_dbus_is_running(Manager *m) {
+        assert(m);
+        return manager_dbus_is_running(Manager *m, false);
```
Comment (yuwata) [inline src/core/manager.h]: Please drop `Manager *`

File: src/core/manager.c [hunk 2]
```diff
@@ -2935,7 +2935,7 @@ static int manager_dispatch_signal_fd(sd_event_source *source, int fd, uint32_t
                 break;
 
         case SIGUSR1:
-                if (manager_dbus_is_running(m, false)) {
+                if (manager_dbus_is_running_full(m, false)) {
```
Comment (yuwata) [inline src/core/manager.c]: Can be `manager_dbus_is_running()`, that is, without `_full`.

File: src/core/manager.c [hunk 3]
```diff
@@ -4140,7 +4140,7 @@ void manager_recheck_dbus(Manager *m) {
         if (MANAGER_IS_RELOADING(m))
                 return; /* don't check while we are reloading… */
 
-        if (manager_dbus_is_running(m, false)) {
+        if (manager_dbus_is_running_full(m, false)) {
```
Comment (yuwata) [inline src/core/manager.c]: Also here.

**Review by yuwata — APPROVED** [review]

File: src/core/dbus-unit.c [hunk 4]
```diff
@@ -1876,6 +1876,12 @@ int bus_unit_queue_job(
             (type == JOB_RELOAD_OR_START && job_type_collapse(type, u) == JOB_START && u->refuse_manual_start))
                 return sd_bus_error_setf(error, BUS_ERROR_ONLY_BY_DEPENDENCY, "Operation refused, unit %s may be requested by dependency only (it is configured to refuse manual start/stop).", u->id);
 
+        /* dbus-broker issues StartUnit for activation requests, so let's apply the same check
+         * used in signal_activation_request(). */
+        if (type == JOB_START && u->type == UNIT_SERVICE &&
+            SERVICE(u)->type == SERVICE_DBUS && !manager_dbus_is_running(m))
```
Comment (yuwata) [inline src/core/dbus-unit.c]: `manager_dbus_is_running(m)` -> `manager_dbus_is_running(u->manager)` ??

Comment (YHNdnzj) [reply src/core/dbus-unit.c]: Oops, sorry. My mind isn't very clear today...

File: src/core/manager.h [hunk 5]
```diff
@@ -504,6 +504,12 @@ DEFINE_TRIVIAL_CLEANUP_FUNC(Manager*, manager_free);
 
 int manager_startup(Manager *m, FILE *serialization, FDSet *fds, const char *root);
 
+bool manager_dbus_is_running_full(Manager *m, bool deserialized);
+static inline bool manager_dbus_is_running(Manager *m) {
+        assert(m);
```
Comment (yuwata) [inline src/core/manager.h]: Not a matter, but we usually do not add `assert()` such case.
(We add assertion e.g. when we access `m->foo` in inline function.)

**Review by yuwata — APPROVED** [review]

**Review by bluca — APPROVED** [review]

**Comment by Geass-LL** [comment]
`manager_dbus_is_running` is not sufficient to me. It doesn't check if there are `stop` pending jobs. If dbus-broker.service is about to shutdown even if it's running now, we should also refuse activating services.

**Comment by YHNdnzj** [comment]
> `manager_dbus_is_running` is not sufficient to me. It doesn't check if there are `stop` pending jobs. If dbus-broker.service is about to shutdown even if it's running now, we should also refuse activating services.

Hmm, @yuwata does it make sense to call `unit_stop_pending` in `manager_dbus_is_running` then?

**Comment by Geass-LL** [comment]
Maybe it's better to use `manager_unit_inactive_or_pending` in both two places. `manager_dbus_is_running` is used in many other places, I'm afraid changing this logic may lead to some unintentional regressions.

**Comment by Geass-LL** [comment]
> Maybe it's better to use `manager_unit_inactive_or_pending` in both two places. `manager_dbus_is_running` is used in many other places, I'm afraid changing this logic may lead to some unintentional regressions.

Forget it. We should't refuse dbus activation if dbus is `inactive`, this will make us fail to start `Type=dbus` services.

Test with this PR merged
```bash
systemctl stop dbus.socket
systemctl start NetworkManager
```
This will fail, and the error message is `Refusing activation, D-Bus is not running`.

**Pull request merged**